src/compiler/checker.ts

```diff
@@ -27970,7 +27970,8 @@ namespace ts {
             // type with those properties for which the binding pattern specifies a default value.
             // If the object literal is spread into another object literal, skip this step and let the top-level object
             // literal handle it instead.
-            if (contextualTypeHasPattern && node.parent.kind !== SyntaxKind.SpreadAssignment) {
+            const spreadOrOutsidePossibleObjectParent = findAncestor(node, n => n.kind === SyntaxKind.SpreadAssignment || isStatement(n) || isArrowFunction(n))!;

```

Comment:
Failing test:

```ts
const { value } = {
  value: "",
  ...(() => true ? {} : {})(),
};
```

I could be missing a subtlety with nested binding patterns, but I think the question we want to answer is, as we are looking up the tree, do we pass a SpreadAssignment before we pass the `contextualType.pattern`? By choosing specific syntax kinds to stop at, you’re essentially trying to write an ad-hoc one-liner to represent what syntaxes contextual types can flow through.

If we wanted to be really sure we arrived at the right answer, we could do something like this:

- walk up the parents, again aborting if we pass the common ancestor of `node` and `contextualType.pattern`
- again stop at a `SpreadAssignment`, but also check to see if its parent’s `getContextualType(parent).pattern === contextualType.pattern`

I’m actually not sure whether it’s possible for that equality check to be false. But if I knew it was true, I’d feel _really_ confident that the check was going to happen on the parent object literal.

Reply:
Related nit: the outer IIFE is totally irrelevant to the repro, so the test name is perhaps a bit misleading.

Reply:
Thanks for the review! I believe that I've addressed the feedback. Could you take another look?